src/aws-cpp-sdk-core/source/client/RequestCompression.cpp

```diff
@@ -0,0 +1,229 @@
+        AWS_REQUEST_COMPRESSION_ALLOCATION_TAG);
+
+    // Prepare ZLIB to compress
+    int ret = Z_NULL;
+    z_stream strm;
+    std::unique_ptr<unsigned char[]> in(new unsigned char[ZLIB_CHUNK]);
+    std::unique_ptr<unsigned char[]> out(new unsigned char[ZLIB_CHUNK]);
+    // Cleaning Zlib stream state flags
+    strm.zalloc = Z_NULL;
+    strm.zfree = Z_NULL;
+    strm.opaque = Z_NULL;
+
+    const int max_window_gzip = 31;
+    const int default_mem_level_usage = 8;
+    ret =
+        deflateInit2(&strm, Z_DEFAULT_COMPRESSION, Z_DEFLATED, max_window_gzip,
+                     default_mem_level_usage, Z_DEFAULT_STRATEGY);
+    if (ret != Z_OK)
+      return false;
+    // Compress
+    //  do {
+    // Fill the buffer
+    if (streamSize > ZLIB_CHUNK) {
+      // Skipping multi chunk for now
+      return false;
+    } else {
+      int flush = Z_FINISH;
+      input->read(reinterpret_cast<char *>(in.get()), streamSize);
+      strm.avail_in = streamSize;
+      strm.next_in = in.get();
+      do {
+        // Run deflate on buffers
+        strm.avail_out = ZLIB_CHUNK;
+        strm.next_out = out.get();
+
+        ret = deflate(&strm, flush);
+
+        // writing the output
+        unsigned output_size = ZLIB_CHUNK - strm.avail_out;
+        output->write(reinterpret_cast<char *>(out.get()), output_size);
+      } while (strm.avail_out == 0);
+    }
+    //  } while (input.length());
+    assert(ret == Z_STREAM_END);
+    AWS_LOGSTREAM_TRACE(AWS_REQUEST_COMPRESSION_LOG_TAG,
+                        "Compressed request to: " << strm.total_out
+                                                  << " bytes");
+    deflateEnd(&strm);
+    return output;
+  } else {
+    AWS_LOGSTREAM_ERROR(
+        AWS_REQUEST_COMPRESSION_LOG_TAG,
+        "Compressed request requested in runtime without support for it");
+    return false;
+  }
+#else
+  // If there is no support to compress, always fail calls to this method.
+  AWS_LOGSTREAM_ERROR(
+      AWS_REQUEST_COMPRESSION_LOG_TAG,
+      "Compressed request requested in runtime without support for it");
+  AWS_UNREFERENCED_PARAM(input);     // silencing warning;
+  AWS_UNREFERENCED_PARAM(algorithm); // silencing warning;
+  return false;
+#endif
+}
+
+Aws::Utils::Outcome<std::shared_ptr<Aws::IOStream>, bool>
+Aws::Client::RequestCompression::uncompress(
+    std::shared_ptr<Aws::IOStream> input,
+    const CompressionAlgorithm &algorithm) const {
+#ifdef ENABLED_ZLIB_REQUEST_COMPRESSION
+  if (algorithm == CompressionAlgorithm::GZIP) {
+    // calculating stream size
+    input->seekg(0, input->end);
+    auto streamSize = input->tellg();
+    input->seekg(0, input->beg);
+
+    AWS_LOGSTREAM_TRACE(AWS_REQUEST_COMPRESSION_LOG_TAG,
+                        "Uncompressing request of " << streamSize << " bytes.");
+
+    // Preparing output
+    std::shared_ptr<Aws::IOStream> output = Aws::MakeShared<Aws::StringStream>(
```

Comment:
this is only a nit now but something I would prefer to enforce
Because of disabled exceptions, we need to check the actual allocated pointer value to handle failed memory allocation cases.

Reply:
Adding